File: src/git_forge_release_migrator/core/checkpoint.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

_TERMINAL_RELEASE_STATUSES = {"created", "updated", "skipped_existing"}
_TERMINAL_TAG_STATUSES = {"tag_created", "tag_skipped_existing"}


def load_checkpoint_state(path: str, signature: str) -> dict[str, str]:
    p = Path(path)
    if not p.exists():
        return {}

    state: dict[str, str] = {}
    with p.open("r", encoding="utf-8") as f:
        for line in f:
            text = line.strip()
            if not text:
                continue
            try:
                item = json.loads(text)
            except json.JSONDecodeError:
                continue
            if not isinstance(item, dict):
                continue
            if str(item.get("signature", "")) != signature:
                continue
            key = str(item.get("key", ""))
            status = str(item.get("status", ""))
            if key and status:
                state[key] = status
    return state


def append_checkpoint(
    path: str,
    *,
    signature: str,
    key: str,
    tag: str,
    status: str,
    message: str,
) -> None:
    record = {
        "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "signature": signature,
        "key": key,
        "tag": tag,
        "status": status,
        "message": message,
    }
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    with p.open("a", encoding="utf-8") as f:
        f.write(json.dumps(record, ensure_ascii=True) + "\n")


def is_terminal_release_status(status: str) -> bool:
    return status in _TERMINAL_RELEASE_STATUSES


def is_terminal_tag_status(status: str) -> bool:
    return status in _TERMINAL_TAG_STATUSES

File: dart_cli/lib/src/core/checkpoint.dart
import 'dart:convert';
import 'dart:io';

import 'time.dart';

const Set<String> _terminalReleaseStatuses = <String>{'created', 'updated', 'skipped_existing'};
const Set<String> _terminalTagStatuses = <String>{'tag_created', 'tag_skipped_existing'};

Map<String, String> loadCheckpointState(String path, String signature) {
  final File file = File(path);
  if (!file.existsSync()) {
    return <String, String>{};
  }

  final Map<String, String> state = <String, String>{};
  final List<String> lines = file.readAsLinesSync();
  for (final String rawLine in lines) {
    final String text = rawLine.trim();
    if (text.isEmpty) {
      continue;
    }

    Map<String, dynamic> item;
    try {
      item = jsonDecode(text) as Map<String, dynamic>;
    } catch (_) {
      continue;
    }

    if ((item['signature'] ?? '').toString() != signature) {
      continue;
    }

    final String key = (item['key'] ?? '').toString();
    final String status = (item['status'] ?? '').toString();
    if (key.isNotEmpty && status.isNotEmpty) {
      state[key] = status;
    }
  }

  return state;
}

void appendCheckpoint(
  String path, {
  required String signature,
  required String key,
  required String tag,
  required String status,
  required String message,
}) {
  final Map<String, String> record = <String, String>{
    'timestamp': utcTimestamp(),
    'signature': signature,
    'key': key,
    'tag': tag,
    'status': status,
    'message': message,
  };

  final File file = File(path);
  file.parent.createSync(recursive: true);
  file.writeAsStringSync('${jsonEncode(record)}\n', mode: FileMode.append);
}

bool isTerminalReleaseStatus(String status) => _terminalReleaseStatuses.contains(status);

bool isTerminalTagStatus(String status) => _terminalTagStatuses.contains(status);
